### ISSUE TITLE
Incomplete writes on Stream

### ISSUE THREAD

**Issue by scalablecory** [dotnet/runtime#48826]
## Background and Motivation

Underlying platform I/O used by `NetworkStream` and `FileStream` already do incomplete writes like this. We emulate complete writes inside of `Stream.Write()`, which may mean multiple syscalls. This will typically happen when the kernel's write buffer is completely filled. See also #48479.

More generalized, a `Stream` implementation may have its own sort of buffering strategy so it may only need to consume part of the input buffer before it must eventually cause a syscall.

Due to this, it's possible a situation happens:

1. The user calls `Write()`.
2. An incomplete write happens.
3. The caller finds additional data that it wants to write.
4. The second write for the remains of the initial `Write()` buffer happens. This write is very small and will have gone out very wastefully: on its own syscall, its own packet, and with its own HTTP/TLS framing, etc.
5. The user calls `Write()` with that additional data.

The result is that we can cause more costly syscalls than are necessary, and possibly less efficient protocol framing. This partially defeats the user's own optimization of having a larger buffer size intending to reduce syscalls.

Having the proposed `WriteSome` APIs would allow us to optimize away step 4. This would reduce CPU usage for high-throughput callers.

## Proposed API

The behavior of this API is "write from a buffer, returning the number of bytes written. in implementations where it is more efficient to do so, the buffer may only be partially written."

The default implementation will be to call `Write()`, writing the entire buffer.

```c#
class Stream
{
    // default implementation: just call Write().
    public virtual int WriteSome(ReadOnlySpan<byte> buffer);
    public virtual ValueTask<int> WriteSomeAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken = default);
}
```

**Comment by stephentoub** [comment]
> The result is that we can cause more costly syscalls than are necessary, and possibly less efficient protocol framing. This partially defeats the user's own optimization of having a larger buffer size intending to reduce syscalls.

Can you share real-world numbers?

Adding a `WriteSomeAsync` overload for example is likely to come with its own deficiencies.  It's default implementation would be something along the lines of:
```
public virtual async ValueTask<int> WriteSomeAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken = default)
{
    await WriteAsync(buffer, cancellationToken);
    return buffer.Length;
}
```
which, in the case where the write completes asynchronously, will make this allocate a whole additional state machine object, plus an extra virtual call, making it measurably less efficient than just using the original WriteAsync directly for any stream that didn't optimize an override.

**Comment by tmds** [comment]
@scalablecory

I think the method is meant to write everything that can be written synchronously. So, `WriteSomeAsync` would never complete async?

> The caller finds additional data that it wants to write.

Why can the caller find additional data at this point?

**Comment by scalablecory** [comment]
> I think the method is meant to write everything that can be written synchronously. So, `WriteSomeAsync` would never complete async?

In the context of sockets, yeah, probably (we should decide if a 0-byte return is reasonable). But in the context of generic `Stream`, I think of this a little differently.

I view a non-full write to indicate that it will be more efficient if the user performs a 2nd write with both a) the leftover data, and b) even more data. One wouldn't really call this if you didn't expect to do more writes later, so that seems like a reasonable definition.

So depending on a stream's goal, something like buffering or enveloping might also impacts its decision as well.

> 
> > The caller finds additional data that it wants to write.
> 
> Why can the caller find additional data at this point?

The two scenarios that got me interested in this feature:

- Forwarding (read from stream A, write to stream B) e.g. YARP.
- In HTTP/2, where you have a lot of concurrency going on within a single stream. It's very possible that, by the time the first I/O completes, I have queued up more requests to be sent out -- do I want to have a 2nd I/O just for the remaining bytes from the previous request, or do I want to have that 2nd I/O also contain all the additional requests?

**Comment by scalablecory** [comment]
@davidfowl I think this this would also be useful for Pipelines, as it's possible the user has written more data. The `SslStream` wrapper being used, as well as the inevitable `Socket` API we'd need to enable this.

**Comment by tmds** [comment]
I think this API moves the responsibility of buffering upstream, and it is missing the notion of backpressure.
A producer should make a couple of `WriteAsync` calls and when it has a meaningful boundary a `FlushAsync`.
These calls should stop the producer from generating more data when the buffer has too much data (by going async).

https://github.com/dotnet/runtime/issues/25344 can help to allow a producer to provide multiple buffers and avoid intermediate flushing.

I think it won't work well either when different producers perform partial writes, because their data would interleave.

**Comment by scalablecory** [comment]
> These calls should stop the producer from generating more data when the buffer has too much data (by going async until data was sent).

I agree, that's why I'm leaning toward a 0-byte write *not* being reasonable. It should always write at least something, and wait if a buffer is full.

But, when the buffer gets a little room, it should stop writing once the buffer is filled again (in the Socket sense, after a single I/O).

> #25344 can help to allow a producer to provide multiple buffers and avoid intermediate flushing.

Yep. It doesn't solve what this issue solves though. I'm using both together.

> I think it won't work well either when different producers perform partial writes, because their data would interleave.

`Stream` makes no guarantees about concurrent writes, so that is not a worry here.

If you're skeptical of my scenarios, here's how I'm implementing HTTP/2 in LLHTTP, and where WriteSome could help:

```c#
Stream stream;
SendQueue buffers;

// thread A.
while(true)
{
    int written = stream.WriteSomeGathered(buffers); // on 2nd loop, there is an opportunity to send out thread C's data too.
    buffers.Discard(written);
}

// thread B.
ReadOnlyMemory<byte> userContentBuffer;

buffers.Add(CreateHttpDataFrameFor(userContentBuffer));
int flushPoint = buffers.Add(userContentBuffer);
await buffers.FlushToAsync(flushPoint); // starts the first call to WriteSomeGathered.

// thread C.
ReadOnlyMemory<byte> headersBuffer, userContentBuffer;

buffers.Add(headersBuffer); // called before the first WriteSomeGathered completes.
int flushPoint = buffers.Add(userContentBuffer);
await buffers.FlushToAsync(flushPoint);
```

**Comment by tmds** [comment]
I see how in the case of a partial send you take it as the opportunity to go back and see if there are more buffers to be sent. I don't think many buffers could be added because `WriteSome` should be fast. Even if there were some, the next send would have to wait for there to be room in the send buffer.

If you'd want to get more buffers, it would be better if you have a way of waiting for the socket to become writable, and then see what buffers are available.

These partial sends happen because the TCP connection can't keep up. I expect that's a bigger bottleneck than the syscalls that WriteSome allows you to save.

**Comment by scalablecory** [comment]
> If you'd want to get more buffers, it would be better if you have a way of waiting for the socket to become writable, and then see what buffers are available.

Possibly. I don't know that. Keeping in mind that this is `Stream` and not `Socket`, do you have an API suggestion I could look at?

> These partial sends happen because the TCP connection can't keep up. I expect that's a bigger bottleneck than the syscalls that `WriteSome` allows you to save.

Indeed :). I expect we can optimize for both -- processes often have more than just one connection. We should definitely measure the perf characteristics under this scenario though.

**Comment by tmds** [comment]
> do you have an API suggestion I could look at?

There's none I know of. We're using zero-buffer reads to check for readable. Zero-buffer writes for writable may be awkward...

**Comment by logiclrd** [comment]
> Adding a `WriteSomeAsync` overload for example is likely to come with its own deficiencies. Its default implementation would be something along the lines of:
> 
> ```
> public virtual async ValueTask<int> WriteSomeAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken = default)
> {
>     await WriteAsync(buffer, cancellationToken);
>     return buffer.Length;
> }
> ```
> 
> which, in the case where the write completes asynchronously, will make this allocate a whole additional state machine object, plus an extra virtual call, making it measurably less efficient than just using the original WriteAsync directly for any stream that didn't optimize an override.

My intuition on this is that the cost of setting up the extra state machine is probably negligible compared to the cost of an additional syscall. Also, how many of these state machines can be set up every second? I wrote up a quick test, and on my system, it seems that approximately 43 million simple `async` state machines can be set up every second. By comparison, a network protocol using this because it expects additional data to be continuously trickling in might be making, what, hundreds? maybe thousands? of `WriteSomeAsync` calls per second. The additional cost should be completely negligible, it would seem. And, there would be absolutely no incentive for a bulk transfer to use any sort of `WriteSome` method. This could be called out in the documentation, too, just tell people: if you're transferring a large amount of data, just use `Write`.
